### PR TITLE
cluster-ui, db-console: fix bug where store ids are treated as node ids

### LINKED ISSUE #118957: ui: db pages are using store ids as node ids for regions mapping
The databases pages display region information for databases and tables using a nodeId -> region map. The value currently being used to lookup region information is not the node id, but the store id. This is because the db details api returns the store ids replicas are located on, not the node ids. This bug likely remained undetected for a while since lots of clusters have 1-1 mapping from store id to node id but store ids and node ids cannot be used interchangeably. Clusters with multiple stores per node will see bugs where missing/incorrect region information is being reported.

See this comment for the table used to return db details on stores being used by the db - https://github.com/cockroachdb/cockroach/blob/43072c2ae08576860c56a7720abd154a4ccf5ed5/pkg/sql/crdb_internal.go#L4451

And here in the UI, where we use node id -> region mappings with the store id - https://github.com/cockroachdb/cockroach/blob/7f8d242b370cfd5a258b688666562db6e5949b8a/pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts#L81-L86

We can either create a store ids -> region map or map each store id to their node and use the existing map. Each node status response should contain the node's stores.

Jira issue: CRDB-35974

### PR DESCRIPTION
**Pull request by laurenbarker**
Previously, store ids were being treated as node ids on the databases view. This resulted in "Regions/Nodes" column showing inaccurate information in some cases. Now, store ids are mapped to node ids on the databases view which resolves the issue.

Fixes: #118957

Release note (ui change): Resolves an issue where clusters with multiple stores per node may list inaccurate region/node information in the databases table.

### CODE REVIEW

File: pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts [hunk 1]
```diff
@@ -147,7 +160,7 @@ const deriveDatabaseTableDetails = (
   const normalizedPrivileges = normalizePrivileges(
     [].concat(...grants.map(grant => grant.privileges)),
   );
-  const nodes = results?.stats.replicaData.nodeIDs || [];
+  const nodes: Nodes = { kind: "node", ids: results?.stats.replicaData.nodeIDs || [] };
```
Comment (laurenbarker) [inline pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts:163]: The table views appear to already be using node IDs. Please correct me if I'm reading this incorrectly.

File: pkg/ui/workspaces/cluster-ui/src/databases/util.tsx [hunk 2]
```diff
@@ -58,31 +58,71 @@ export class GetDatabaseInfoError extends Error {
   }
 }
 
-// getNodesByRegionString converts a list of node ids and map of
-// node ids to region to a string of node ids by region, ordered
-// by region name, e.g.
-// regionA(n1, n2), regionB(n2,n3), ...
+/** Store ids and node ids are both of type `number[]`. To disambiguate, a
+ * `kind` field is included in the type. */
+export type Stores = { kind: "store", ids: number[] };
```
Comment (laurenbarker) [inline pkg/ui/workspaces/cluster-ui/src/databases/util.tsx:63]: Expanding the type is the only way I could think to disambiguate the data of type `number[]`... Open to alternatives here.

File: pkg/ui/workspaces/cluster-ui/src/databases/util.tsx [hunk 3]
```diff
@@ -102,7 +142,7 @@ export function nodesByRegionMapToString(
     .join(", ");
 }
 
-export function createNodesByRegionMap(
+function createNodesByRegionMap(
```
Comment (laurenbarker) [inline pkg/ui/workspaces/cluster-ui/src/databases/util.tsx:145]: `nodesByRegionMapToString` and `createNodesByRegionMap` are only used in this file, so I removed the exports.

File: pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts [hunk 4]
```diff
@@ -164,7 +177,7 @@ const deriveDatabaseTableDetails = (
       statsLastUpdated: results?.heuristicsDetails,
       indexStatRecs: results?.stats.indexStats,
       spanStats: results?.stats.spanStats,
-      nodes: nodes,
+      nodes: nodes.ids,
```
Comment (laurenbarker) [inline pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts:180]: I'm assuming this should be node ids and not store ids too.

File: pkg/ui/workspaces/cluster-ui/src/databasesPage/databasesPage.tsx [hunk 5]
```diff
@@ -59,6 +59,7 @@ import {
 } from "../api";
 import { InlineAlert } from "@cockroachlabs/ui-components";
 import { checkInfoAvailable } from "../databases";
+import { cockroach } from "../../../../../../_bazel/bin/pkg/ui/workspaces/db-console/src/js/protos";
```
Comment (laurenbarker) [inline pkg/ui/workspaces/cluster-ui/src/databasesPage/databasesPage.tsx:62]: Oops, seems like this is not the correct import path. Fixing.

**Comment by laurenbarker** [comment]
Looks like I need to stub an API response in `pkg/ui/workspaces/db-console/src/views/databases/databasesPage/redux.spec.ts`. I'll do that tomorrow.

**Comment by laurenbarker** [comment]
Rebased on changes from https://github.com/cockroachdb/cockroach/pull/118996 was merged.

**Comment by xinhaoz** [comment]
*[`pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts` line 181 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119260#-NrCR-KtDhs8ew0agTRz:-NrCR-KtDhs8ew0agTS-:b-87mdbg) ([raw file](https://github.com/cockroachdb/cockroach/blob/68fe9628369179ada456ca7fbad72e7d0c54a443/pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts#L181)):*
> ```XML
>   );
>   /** List of store IDs for the current cluster. All of the values in the
>    * `*replicas` columns correspond to store IDs. */
> ```

nit: I think this comment can be deleted now since the field was renamed.

**Review by xinhaoz — COMMENTED** [review]
Thanks so much for taking this on (and taking the time to understand what's going on with db console as a whole :)), just some super small nits but otherwise all good :100: 

> Reviewed 3 of 7 files at r1, 7 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119260#-:-NrCW_1A9R6bdV2piEwY:b-v1m92r)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @laurenbarker)

___
*[`pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts` line 163 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119260#gh-1491656146:-NrCTeZ130EM96jMIKuk:b-zbxvm) ([raw file](https://github.com/cockroachdb/cockroach/blob/b83f7543f02a6cc06d9e00fb74ee0f6c10fe5e0c/pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts#L163)):*
<details><summary><i>Previously, laurenbarker (Lauren Barker) wrote…</i></summary><blockquote>

The table views appear to already be using node IDs. Please correct me if I'm reading this incorrectly.
</blockquote></details>

Yeah, I believe all the database page views assumed they were using node ids and store ids.
___
*[`pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts` line 180 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119260#gh-1491658831:-NrCSPWA4ZeCSvNcKsdJ:b40qle2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b83f7543f02a6cc06d9e00fb74ee0f6c10fe5e0c/pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts#L180)):*
<details><summary><i>Previously, laurenbarker (Lauren Barker) wrote…</i></summary><blockquote>

I'm assuming this should be node ids and not store ids too.
</blockquote></details>

yup! :thumbsup: 
___
*[`pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts` line 241 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119260#-NrCU5w9AFiECCrRpiNA:-NrCU5w9AFiECCrRpiNB:b8qeh3l) ([raw file](https://github.com/cockroachdb/cockroach/blob/68fe9628369179ada456ca7fbad72e7d0c54a443/pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts#L241)):*
> ```XML
> 
>     /** List of store IDs for the current cluster. All of the values in the
>      * `*replicas` columns correspond to store IDs. */
> ```

same here with the comment
___
*[`pkg/ui/workspaces/cluster-ui/src/databases/util.tsx` line 145 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119260#gh-1491658086:-NrCUlRJ0IJUe6YRTHd2:bsv417h) ([raw file](https://github.com/cockroachdb/cockroach/blob/b83f7543f02a6cc06d9e00fb74ee0f6c10fe5e0c/pkg/ui/workspaces/cluster-ui/src/databases/util.tsx#L145)):*
<details><summary><i>Previously, laurenbarker (Lauren Barker) wrote…</i></summary><blockquote>

`nodesByRegionMapToString` and `createNodesByRegionMap` are only used in this file, so I removed the exports.
</blockquote></details>

thank you!
___
*[`pkg/ui/workspaces/cluster-ui/src/databases/util.tsx` line 71 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119260#-NrCWIyR2_6P4XemvgV5:-NrCWIyR2_6P4XemvgV6:b-3f7qo9) ([raw file](https://github.com/cockroachdb/cockroach/blob/68fe9628369179ada456ca7fbad72e7d0c54a443/pkg/ui/workspaces/cluster-ui/src/databases/util.tsx#L71)):*
> ```XML
>  * @returns A list of node ids for the cluster.
>  */
> export function getNodeIdsFromStoreIds(
> ```

Can we just make this return the `Nodes` type directly since all its use cases seem to be creating Nodes?
___
*[`pkg/ui/workspaces/cluster-ui/src/databasesPage/databasesPage.tsx` line 112 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119260#-NrCVbwZ8H1kpNTynDL6:-NrCVbwZ8H1kpNTynDL7:b-kfgyvy) ([raw file](https://github.com/cockroachdb/cockroach/blob/68fe9628369179ada456ca7fbad72e7d0c54a443/pkg/ui/workspaces/cluster-ui/src/databasesPage/databasesPage.tsx#L112)):*
> ```XML
>   nodes?: number[];
>   /** A list of node statuses so that store ids can be mapped to nodes. */
>   nodeStatuses: cockroach.server.status.statuspb.INodeStatus[];
> ```

I don't think we need to add this to the props.

**Comment by xinhaoz** [comment]
*[`pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts` line 163 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119260#gh-1491656146:-NrCZIW69Y5HnePA50zd:bd5t6rr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b83f7543f02a6cc06d9e00fb74ee0f6c10fe5e0c/pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts#L163)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Yeah, I believe all the database page views assumed they were using node ids and store ids.
</blockquote></details>

not store ids**

**Review by laurenbarker — COMMENTED** [review]
TFTR @xinhaoz! I think I need help with the last test -- "loads table data" in `pkg/ui/workspaces/db-console/src/views/databases/databaseDetailsPage/redux.spec.ts`. `nodeStatuses` is `undefined` and I'm not sure how to get that information populated in the store and if I need to modify non-test code as well.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119260#-:-NrCvCDt1nXVzLGEwGGL:b-g3pmzf)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @xinhaoz)

___
*[`pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts` line 163 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119260#gh-1491656146:-NrCdxI8Bd24hK-X1Wmp:b-3mlkyo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b83f7543f02a6cc06d9e00fb74ee0f6c10fe5e0c/pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts#L163)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

not store ids**
</blockquote></details>

Oops, this was a comment from before you made your change. I wasn't sure if `replicaData.nodeIDs` were node ids or store ids. Turns out they were store ids and you updated the name :smile: 
___
*[`pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts` line 181 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119260#-NrCR-KtDhs8ew0agTRz:-NrCeDqODnhX7V6uc0ze:bombmv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/68fe9628369179ada456ca7fbad72e7d0c54a443/pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts#L181)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

nit: I think this comment can be deleted now since the field was renamed.
</blockquote></details>

Good call. Done.
___
*[`pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts` line 241 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119260#-NrCU5w9AFiECCrRpiNA:-NrCeb-T3Gsmt368PU8F:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/68fe9628369179ada456ca7fbad72e7d0c54a443/pkg/ui/workspaces/cluster-ui/src/databases/combiners.ts#L241)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

same here with the comment
</blockquote></details>

Done.
___
*[`pkg/ui/workspaces/cluster-ui/src/databases/util.tsx` line 71 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119260#-NrCWIyR2_6P4XemvgV5:-NrCfNDnAAiCeGiZMHjX:b-qbtrg9) ([raw file](https://github.com/cockroachdb/cockroach/blob/68fe9628369179ada456ca7fbad72e7d0c54a443/pkg/ui/workspaces/cluster-ui/src/databases/util.tsx#L71)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Can we just make this return the `Nodes` type directly since all its use cases seem to be creating Nodes?
</blockquote></details>

Oh yeah, let's do that!
___
*[`pkg/ui/workspaces/cluster-ui/src/databasesPage/databasesPage.tsx` line 112 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119260#-NrCVbwZ8H1kpNTynDL6:-NrCgQ56BwG_e0Tg-BJU:b-xrerlv) ([raw file](https://github.com/cockroachdb/cockroach/blob/68fe9628369179ada456ca7fbad72e7d0c54a443/pkg/ui/workspaces/cluster-ui/src/databasesPage/databasesPage.tsx#L112)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

I don't think we need to add this to the props.
</blockquote></details>

Oops, removed!

**Review by xinhaoz — APPROVED** [review]

**Comment by laurenbarker** [comment]
TFTRs @xinhaoz!

**Comment by laurenbarker** [comment]
blathers backport 23.2 23.1

**Pull request merged**